fapolicy_analyzer/ui/features/profiler_feature.py

```diff
@@ -0,0 +1,175 @@
+
+gi.require_version("Gtk", "3.0")
+from gi.repository import GLib  # isort: skip
+
+PROFILING_FEATURE = "profiling"
+_handle: ProcHandle
+
+
+def create_profiler_feature(dispatch: Callable) -> ReduxFeatureModule:
+    profiler_active: bool = False
+
+    def _idle_dispatch(action: Action):
+        GLib.idle_add(dispatch, action)
+
+    def _init_profiler() -> Action:
+        return profiler_init()
+
+    def _on_exec(h: ExecHandle, action_fn: Callable[[int], Action]):
+        _idle_dispatch(action_fn(h.pid))
+        _idle_dispatch(set_profiler_output(h.event_log, h.stdout_log, h.stderr_log))
+
+    def _on_tick(_: ExecHandle, duration: int, action_fn: Callable[[int], Action]):
+        _idle_dispatch(action_fn(duration))
+
+    def _on_done(action_fn: Callable[[], Action], flag_fn: Callable[[], None]):
+        _idle_dispatch(action_fn())
+        flag_fn()
+
+    def _start_profiling(action: Action) -> Action:
+        global _handle
+        nonlocal profiler_active
+
+        def on_done():
+            nonlocal profiler_active
+            profiler_active = False
+
+        if profiler_active:
+            return action
+
+        profiler_active = True
+
+        exed = partial(
+            _on_exec,
+            action_fn=profiler_exec,
+        )
+
+        tick = partial(
+            _on_tick,
+            action_fn=profiler_tick,
+        )
+
+        done = partial(
+            _on_done,
+            action_fn=profiler_done,
+            flag_fn=on_done,
+        )
+
+        # build target command from session args
+        args: Dict[str, str] = action.payload
+        target = args.get("cmd", None)
+        target_args = args.get("arg", None)
+        cmd = target if target else None
+        if cmd and target_args:
+            cmd = f"{target} {target_args}"
+
+        # configure the backend
+        try:
+            p = Profiler()
+
+            # set callbacks
+            p.exec_callback = exed
+            p.tick_callback = tick
+            p.done_callback = done
+
+            # set user, pwd, envs
+            p.user = args.get("uid", None)
+            p.pwd = args.get("pwd", None)
+            p.env = args.get("env", None)
+        except RuntimeError:
+            dispatch(profiler_initialization_error(PROFILER_INIT_ERROR))
+            return profiler_done()
+
+        # execute and dispatch
+        try:
+            _handle = p.profile(cmd)
+        except RuntimeError:
+            dispatch(profiler_execution_error(PROFILER_EXEC_ERROR))
+            return profiler_done()
+
+        return profiling_started(cmd)
+
+    def _kill_profiler(action: Action) -> Action:
+        global _handle
+        nonlocal profiler_active
+
+        if not profiler_active:
+            return action
+
+        _handle.kill()
+
+        return terminating_profiler()
+
+    init_epic = pipe(
+        of_init_feature(PROFILING_FEATURE),
+        map(lambda _: _init_profiler()),
+    )
```

Comment:
I don't think you need this. The epic here isn't doing any work that would change the state. I think it final state after this one would just be the default state of the `profile_reducer` correct?

Reply:
Yep, that's right :+1: